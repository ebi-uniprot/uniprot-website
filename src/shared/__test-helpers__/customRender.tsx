/**
 * This tries to follow what is detailed here:
 * https://testing-library.com/docs/react-testing-library/setup#custom-render
 */
import { render, RenderOptions } from '@testing-library/react';
import { Component, Dispatch, ReactElement, ReactNode } from 'react';
import { HelmetProvider } from 'react-helmet-async';
import {
  createMemoryHistory,
  LocationState,
  MemoryHistory,
  Route,
  Router,
} from 'react-router';
import { JsonValue, SetRequired } from 'type-fest';

import { MessagesState } from '../../messages/state/messagesInitialState';
import { MessagesAction } from '../../messages/state/messagesReducers';
import databaseInfo from '../../uniprotkb/utils/__tests__/__mocks__/databaseInfo';
import { getDatabaseInfoMaps } from '../../uniprotkb/utils/database';
import {
  MessagesDispatchContext,
  MessagesStateContext,
} from '../contexts/Messages';
import {
  DatabaseInfoMapsContext,
  UniProtDataVersionContext,
} from '../contexts/UniProtData';
import { JobAction } from '../workers/jobs/state/actionHandler';
import { JobsState } from '../workers/jobs/state/jobsInitialState';

type ExtraRenderOptions = {
  // For react-router
  history?: MemoryHistory<LocationState>;
  path?: string;
  // For custom user preferences (used by useLocalStorage)
  initialLocalStorage?: Record<string, JsonValue>;
  // For app context
  messagesState?: MessagesState;
  messagesDispatch?: jest.Mock<Dispatch<MessagesAction>>;
  JobsState?: JobsState;
  toolsDispatch?: jest.Mock<Dispatch<JobAction>>;
};

type WrapperProps = { children: ReactNode } & RenderOptions &
  SetRequired<
    ExtraRenderOptions,
    | 'history'
    | 'initialLocalStorage'
    | 'messagesState'
    | 'messagesDispatch'
    | 'JobsState'
    | 'toolsDispatch'
  >;

const dbInfoMaps = getDatabaseInfoMaps(databaseInfo);
const uniProtDataVersion = {
  releaseNumber: '2023_01',
  releaseDate: new Date('01-March-2023'),
};

class Wrapper extends Component<WrapperProps> {
  constructor(props: WrapperProps) {
    super(props);
    for (const [key, value] of Object.entries(props.initialLocalStorage)) {
      window.localStorage.setItem(key, JSON.stringify(value));
    }
  }

  // eslint-disable-next-line class-methods-use-this
  componentWillUnmount() {
    window.localStorage.clear();
  }

  render() {
    const { children, path, history, messagesDispatch, messagesState } =
      this.props;
    return (
      <HelmetProvider>
        <Router history={history}>
          <MessagesDispatchContext.Provider value={messagesDispatch}>
            <MessagesStateContext.Provider value={messagesState}>
              <DatabaseInfoMapsContext.Provider value={dbInfoMaps}>
                <UniProtDataVersionContext.Provider value={uniProtDataVersion}>
                  {path ? (
                    <Route path={path} render={() => children} />
                  ) : (
                    children
                  )}
                </UniProtDataVersionContext.Provider>
              </DatabaseInfoMapsContext.Provider>
            </MessagesStateContext.Provider>
          </MessagesDispatchContext.Provider>
        </Router>
      </HelmetProvider>
    );
  }
}

const customRender = (
  ui: ReactElement,
  {
    route = '',
    path,
    history = createMemoryHistory({ initialEntries: [route] }),
    initialLocalStorage = {},
    messagesState = {},
    messagesDispatch = jest.fn(),
    JobsState = {},
    toolsDispatch = jest.fn(),
    ...options
  }: RenderOptions &
    ExtraRenderOptions & {
      route?: string;
    } = {}
) => ({
  ...render(ui, {
    wrapper: (props) => (
      <Wrapper
        path={path}
        history={history}
        initialLocalStorage={initialLocalStorage}
        messagesState={messagesState}
        messagesDispatch={messagesDispatch}
        JobsState={JobsState}
        toolsDispatch={toolsDispatch}
        {...props}
      />
    ),
    ...options,
  }),
  history,
  messagesDispatch,
  toolsDispatch,
});

export default customRender;
