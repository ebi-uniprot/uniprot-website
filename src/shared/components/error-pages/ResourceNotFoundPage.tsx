import { Redirect, useLocation } from 'react-router-dom';
import { Message } from 'franklin-sites';

import ErrorPage from './ErrorPage';

import ArtWork from './svgs/404.svg';
import { Namespace } from '../../types/namespaces';

// Regular expression magic incantations ðŸª„
const redirectMap = new Map<RegExp, string>([
  // main data
  [/^\/uniprot(?<rest>\/.*)?$/i, `/${Namespace.uniprotkb}$<rest>`],
  [/^\/unipark(?<rest>\/.*)?$/i, `/${Namespace.uniparc}$<rest>`],
  [/^\/proteome(?<rest>\/.*)?$/i, `/${Namespace.proteomes}$<rest>`],
  // supporting data
  [/^\/keyword(?<rest>\/.*)?$/i, `/${Namespace.keywords}$<rest>`],
  [
    /^\/(citation|literatures?|publications?|papers?)(?<rest>\/.*)?$/i,
    `/${Namespace.citations}$<rest>`,
  ],
  [/^\/disease(?<rest>\/.*)?$/i, `/${Namespace.diseases}$<rest>`],
  // exception, this one needs to be singular
  [
    /^\/(((cross|x)[-_ ]?)?referenced?[-_ ]?)?databases(?<rest>\/.*)?$/i,
    `/${Namespace.database}$<rest>`,
  ],
  [
    /^\/((sub[-_ ]?)?cellular[-_ ]?)?location(?<rest>\/.*)?$/i,
    `/${Namespace.locations}$<rest>`,
  ],
  // tools
  [/^\/(tools-?)?dashboard(?<rest>\/.*)?$/i, `/tool-dashboard$<rest>`],
  [/^\/peptidesearch(?<rest>\/.*)?$/i, `/peptide-search$<rest>`],
  // TODO: check final URL for those
  [/^\/upload-?lists?(?<rest>\/.*)?$/i, `/id-mapping$<rest>`],
]);

// eslint-disable-next-line consistent-return
export const redirectFromTo = (from: string) => {
  for (const [pattern, replacement] of redirectMap.entries()) {
    if (pattern.test(from)) {
      return from.replace(pattern, replacement);
    }
  }
};

const ErrorMessage = () => (
  <Message level="failure">
    <h4>Sorry, this page can&apos;t be found!</h4>
    <span>Please check the address bar for any mistakes</span>
  </Message>
);

const ResourceNotFoundPage = () => {
  const location = useLocation();

  const newPathname = redirectFromTo(location.pathname);

  if (newPathname) {
    let state = { redirectFrom: location.pathname };
    if (location.state && typeof location.state === 'object') {
      state = { ...location.state, ...state };
    }
    return <Redirect to={{ ...location, pathname: newPathname, state }} />;
  }

  return (
    <ErrorPage
      artwork={<ArtWork />}
      message={<ErrorMessage />}
      data-testid="error-page"
    />
  );
};

export default ResourceNotFoundPage;
