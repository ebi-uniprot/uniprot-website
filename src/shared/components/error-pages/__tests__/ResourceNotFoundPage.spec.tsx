import { screen } from '@testing-library/react';

import customRender from '../../../__test-helpers__/customRender';

import ResourceNotFoundPage, { redirectFromTo } from '../ResourceNotFoundPage';

describe('ResourceNotFoundPage component', () => {
  test('should render', () => {
    const { asFragment, history } = customRender(<ResourceNotFoundPage />, {
      route: '/uniprotkb',
    });
    expect(history.location.pathname).toBe('/uniprotkb');
    expect(screen.getByTestId('error-page')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  test('should redirect to correct page', () => {
    const { history } = customRender(<ResourceNotFoundPage />, {
      route: '/unipark/UPI01',
    });
    expect(history.location.pathname).toBe('/uniparc/UPI01');
    expect(history.location.state).toEqual(
      expect.objectContaining({
        redirectFrom: '/unipark/UPI01',
      })
    );
  });
});

const correctURLs = new Set([
  '/keywords',
  '/uniparc/UPI01',
  '/uniprotkb/P00001/entry',
]);

const incorrectURLs = new Map<string, string>([
  ['/uniprot', '/uniprotkb'],
  ['/uniprot/P00001', '/uniprotkb/P00001'],
  ['/cross-referenced databases', '/database'],
  ['/publication', '/citations'],
]);

describe('replacement patterns', () => {
  test('should not provide a redirection', () => {
    for (const correct of correctURLs) {
      expect(redirectFromTo(correct)).toBeUndefined();
    }
  });

  test('should provide a redirection', () => {
    for (const [incorrect, correct] of incorrectURLs) {
      expect(redirectFromTo(incorrect)).toBe(correct);
    }
  });
});
