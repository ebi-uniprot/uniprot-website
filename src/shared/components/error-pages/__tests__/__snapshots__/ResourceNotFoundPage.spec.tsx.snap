// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResourceNotFoundPage component should render 1`] = `
<DocumentFragment>
  <div
    class="error-page-container"
    data-testid=""
  >
    <test-file-stub
      classname="error-page-container__art-work"
    />
    <div
      class="message message--failure"
      role="status"
    >
      <div
        class="message__side-border"
      />
      <test-file-stub
        height="1.125em"
        width="1.125em"
      />
      <section
        class="message__content"
      >
        <small>
          <h4>
            Sorry, this page can't be found!
          </h4>
          <span>
            Please check the address bar for any mistakes
          </span>
        </small>
      </section>
    </div>
  </div>
</DocumentFragment>
`;
